src/shared/varlink-idl.c

```diff
@@ -0,0 +1,1586 @@
+                return varlink_idl_validate_symbol(field->symbol, v, VARLINK_REGULAR, NULL);
+
+        case VARLINK_BOOL:
+                if (!json_variant_is_boolean(v))
+                        return log_debug_errno(SYNTHETIC_ERRNO(EMEDIUMTYPE), "Field '%s' should be a bool, but it is not, refusing.", strna(field->name));
+
+                break;
+
+        case VARLINK_INT:
+                if (!json_variant_is_integer(v) && !json_variant_is_unsigned(v))
+                        return log_debug_errno(SYNTHETIC_ERRNO(EMEDIUMTYPE), "Field '%s' should be an int, but it is not, refusing.", strna(field->name));
+
+                break;
+
+        case VARLINK_FLOAT:
+                if (!json_variant_is_number(v))
+                        return log_debug_errno(SYNTHETIC_ERRNO(EMEDIUMTYPE), "Field '%s' should be a float, but it is not, refusing.", strna(field->name));
+
+                break;
+
+        case VARLINK_STRING:
+                if (!json_variant_is_string(v))
+                        return log_debug_errno(SYNTHETIC_ERRNO(EMEDIUMTYPE), "Field '%s' should be a string, but it is not, refusing.", strna(field->name));
+
+                break;
+
+        case VARLINK_OBJECT:
+                if (!json_variant_is_object(v))
+                        return log_debug_errno(SYNTHETIC_ERRNO(EMEDIUMTYPE), "Field '%s' should be an object, but it is not, refusing.", strna(field->name));
+
+                break;
+
+        default:
+                assert_not_reached();
+        }
+
+        return 0;
+}
+
+static int varlink_idl_validate_field(const VarlinkField *field, JsonVariant *v) {
+        int r;
+
+        assert(field);
+
+        if (!v || json_variant_is_null(v)) {
+
+                if (!FLAGS_SET(field->field_flags, VARLINK_NULLABLE))
+                        return log_debug_errno(SYNTHETIC_ERRNO(ENOANO), "Mandatory field '%s' is null or missing on object, refusing.", strna(field->name));
+
+        } else if (FLAGS_SET(field->field_flags, VARLINK_ARRAY)) {
+                JsonVariant *i;
+
+                if (!json_variant_is_array(v))
+                        return log_debug_errno(SYNTHETIC_ERRNO(EMEDIUMTYPE), "Field '%s' should be an array, but it is not, refusing.", strna(field->name));
+
+                JSON_VARIANT_ARRAY_FOREACH(i, v) {
+                        r = varlink_idl_validate_field_element_type(field, i);
+                        if (r < 0)
+                                return r;
+                }
+
+        } else if (FLAGS_SET(field->field_flags, VARLINK_MAP)) {
+                const char *k;
+                JsonVariant *e;
+
+                if (!json_variant_is_object(v))
+                        return log_debug_errno(SYNTHETIC_ERRNO(EMEDIUMTYPE), "Field '%s' should be an object, but it is not, refusing.", strna(field->name));
+
+                JSON_VARIANT_OBJECT_FOREACH(k, e, v) {
+                        (void) k; /* Suppress unused warnings from compiler */
```

Comment:
Interesting. Maybe, better to use `_unused_` attribute instead? Though, I am not sure if it works.